presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -169,18 +174,24 @@ public Page getOutput()
                 }
                 currentProbePositionProducedRow = false;
                 if (!advanceProbePosition()) {
+                    probeEnd = true;
                     break;
                 }
                 statisticsCounter.recordProbe(joinSourcePositions);
                 joinSourcePositions = 0;
             }
         }
 
-        // only flush full pages unless we are done
-        if (pageBuilder.isFull() || (finishing && !pageBuilder.isEmpty() && probe == null)) {
-            Page page = pageBuilder.build();
+        // flush a page if we are done, or
+        // when probe reached the end; indices cannot span multiple probe pages.
+        if (pageBuilder.isFull() || probeEnd || (finishing && !pageBuilder.isEmpty())) {
```

Comment:
`(finishing && !pageBuilder.isEmpty())` used to contain a clause equivalent to `&& probeEnd`. Without it, this is a wrong condition "when input has been marked as finished (i.e. this is last probe), let's produce tiny pages".

In fact, when you force-flush at every probe end, the `|| (finishing && !pageBuilder.isEmpty())` can be removed.